File: Week1/#1_First_Bad_Version/Solution.js
/*
 * Created on Fri May 01 2020
 *
 * Title: Leetcode - First Bad Version
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

 /**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        if(n < 2) {
            return n;
        }
        let start = 1;
        let end = n;
        while( start <= end ) {
            let mid = Math.trunc((start + end) / 2);
            if(isBadVersion(mid) && !isBadVersion(mid - 1)) {
                return mid;
            }
            else if(isBadVersion(mid - 1)) {
                end = mid - 1;
            }
            else {
                start = mid + 1;
            }
        }
        return -1;
    };
};

File: Week1/#1_First_Bad_Version/Solution.cpp
/*
 * Created on Fri May 01 2020
 *
 * Title: Leetcode - First Bad Version
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

#include <iostream>

using namespace std;

// The API isBadVersion is defined for you.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        if(n < 2) {
            return n;
        }

        int start_val = 1;
        int end_val = n;

        while(start_val <= end_val) {
            int mid_val = start_val + (end_val - start_val) / 2;
            if(isBadVersion(mid_val) && !isBadVersion(mid_val - 1)) {
                return mid_val;
            }
            else if(isBadVersion(mid_val - 1)) {
                end_val = mid_val - 1;
            }
            else {
                start_val = mid_val +1;
            }
        }

        return -1; // in case if no bad version is found, otherwise compiler will through warning.
    }
};